logger/facility.go

```diff
@@ -0,0 +1,211 @@
+	// No level interval check is needed because level values can't be
+	// created from outside of this package.
+	l := level.Level()
+	lf.levels[name] = l
+	return nil
+}
+
+func (lf *loggerFacility) SetLogger(logger bark.Logger) {
+	lf.logger = logger
+}
+
+func (lf *loggerFacility) Logger(name string) Logger {
+	return &namedLogger{
+		name: name,
+		recv: lf,
+	}
+}
+
+func (lf *loggerFacility) WithField(key string, value interface{}) Factory {
+	return &loggerFacilityFields{
+		recv:   lf,
+		fields: Fields{key: value},
+	}
+}
+
+func (lf *loggerFacility) WithFields(fields Fields) Factory {
+	return &loggerFacilityFields{
+		recv:   lf,
+		fields: fields,
+	}
+}
+
+func (lf *loggerFacility) shouldOutput(name string, level level) bool {
+	minLevel := defaultNotConfiguredNamedLogger
+	if confLevel, ok := lf.levels[name]; ok {
+		minLevel = confLevel
+	}
+	return minLevel <= level
+}
+
+func (lf *loggerFacility) getLogger(fields Fields) bark.Logger {
+	logger := lf.logger
+	if len(fields) > 0 {
+		logger = logger.WithFields(fields)
+	}
+	return logger
+}
+
+type namedMsg struct {
+	name   string
+	level  level
+	fields Fields
+	args   []interface{}
+}
+
+// Having this method makes updating configuration convenient.
+// Without it, a call on SetLevel must push an update to all namedLogger
+// instances. The bookkeeping of all instances is not trivial because any
+// namedLogger can create additional instances when WithField/WithFiels is
+// called. All those references must be tracked carefully using indirect
+// references to allow the GC to reclaim memory.
+// Making all namedLogger instances delegate back to this method sidesteps
+// this problem entirely.
+func (lf *loggerFacility) Log(msg *namedMsg) {
+	if !lf.shouldOutput(msg.name, msg.level) {
+		return
+	}
+	logger := lf.getLogger(msg.fields)
+	switch msg.level {
+	case Trace:
+		logger.Debug(msg.args...)
+	case Debug:
+		logger.Debug(msg.args...)
+	case Info:
+		logger.Info(msg.args...)
+	case Warn:
+		logger.Warn(msg.args...)
+	case Error:
+		logger.Error(msg.args...)
+	case Fatal:
+		logger.Fatal(msg.args...)
+	case Panic:
+		logger.Panic(msg.args...)
+	}
+}
+
+type namedfMsg struct {
+	*namedMsg
+	format string
+}
+
+func (lf *loggerFacility) Logf(msg *namedfMsg) {
+	if !lf.shouldOutput(msg.name, msg.level) {
+		return
+	}
+	logger := lf.getLogger(msg.fields)
+	switch msg.level {
+	case Trace:
+		logger.Debugf(msg.format, msg.args...)
+	case Debug:
+		logger.Debugf(msg.format, msg.args...)
+	case Info:
+		logger.Infof(msg.format, msg.args...)
+	case Warn:
+		logger.Warnf(msg.format, msg.args...)
+	case Error:
+		logger.Errorf(msg.format, msg.args...)
+	case Fatal:
+		logger.Fatalf(msg.format, msg.args...)
+	case Panic:
+		logger.Panicf(msg.format, msg.args...)
+	}
+}
+
+type logReceiver interface {
+	Log(*namedMsg)
+	Logf(*namedfMsg)
+}
+
+type loggerFacilityFields struct {
+	recv   logReceiver
+	fields Fields
+}
+
+func (lff *loggerFacilityFields) Log(msg *namedMsg) {
```

Comment:
This is a good point. `namedMsg` is just a plain struct used move data. We can change `Log`'s signature to accept same data a `namedMsg` carries as arguments.